projects/plugins/jetpack/modules/masterbar/admin-menu/class-base-admin-menu.php

```diff
@@ -221,6 +257,71 @@ public function enqueue_scripts() {
 		);
 	}
 
+	/**
+	 * Hide the submenu page based on slug and return the item that was hidden.
+	 *
+	 * Instead of actually removing the submenu item, a safer approach is to hide it and filter it in the API response.
+	 * In this manner we'll avoid breaking third-party plugins depending on items that no longer exist.
+	 *
+	 * A false|array value is returned to be consistent with remove_submenu_page() function
+	 *
+	 * @param string $menu_slug The parent menu slug.
+	 * @param string $submenu_slug The submenu slug that should be hidden.
+	 * @return false|array
+	 */
+	public function hide_submenu_page( $menu_slug, $submenu_slug ) {
+		global $submenu;
+
+		if ( ! isset( $submenu[ $menu_slug ] ) ) {
+			return false;
+		}
+
+		foreach ( $submenu[ $menu_slug ] as $i => $item ) {
+			if ( $submenu_slug !== $item[2] ) {
+				continue;
+			}
+
+			$this->hide_submenu_element( $i, $menu_slug, $item );
+
+			return $item;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Apply the hide-if-js and hide-if-no-js CSS classes to a submenu item.
+	 *
+	 * @param int    $index The position of a submenu item in the submenu array.
+	 * @param string $parent_slug The parent slug.
+	 * @param array  $item The submenu item.
+	 */
+	public function hide_submenu_element( $index, $parent_slug, $item ) {
+		global $submenu;
+
+		$css_classes = empty( $item[4] ) ? self::HIDE_CSS_CLASS : $item[4] . ' ' . self::HIDE_CSS_CLASS;
+
+		// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+		$submenu [ $parent_slug ][ $index ][4] = $css_classes;
+	}
+
+	/**
+	 * Check if the menu has submenu items visible
+	 *
+	 * @param array $submenu_items The submenu items.
+	 * @return bool
+	 */
+	public function has_visible_items( $submenu_items ) {
+		$visible_items = array_filter(
+			$submenu_items,
+			static function ( $item ) {
+				return empty( $item[4] ) || strpos( $item[4], self::HIDE_CSS_CLASS ) === false;
```

Comment:
I think this doesn't exclude invisible menu items having a class like `my-menu-class hide-if-js` (this pattern is used in plugins like [WooCommerce](https://github.com/woocommerce/woocommerce-admin/blob/07d26b12847647b273db2c285315b7dfa8f550ec/src/Features/Navigation/CoreMenu.php#L292-L299)):

```
strpos( 'my-menu-class hide-if-js', 'hide-if-js hide-if-no-js' ) // false 
```

Reply:
Hmm... I should look into it, thanks for bringing this up! :)

Reply:
Since we are now hiding the elements just with `hide-if-js` class, I don't think we need any other changes for this. :)